File: _experimental/_r/numbers/src/numbers_iterative.R
fibonacci <- function(n) {
  return(fibonacci_iter(n, 0, 1))
}

fibonacci_iter <- function(n, acc2, acc1) {
  if (n <= 0) {
    return(acc2)
  } else if (n <= 2) {
    return(acc1 + acc2)
  } else {
    return(fibonacci_iter(n - 1, acc1, acc1 + acc2))
  }
}

factorial <- function(n) {
  return(factorrial_iter(n, 1))
}

factorrial_iter <- function(n, acc) {
  if (n <= 1) {
    return(acc)
  } else {
    return(factorrial_iter(n - 1, n * acc))
  }
}

sum_numbers <- function(n) {
  return(sum_numbers_iter(n, 0))
}

sum_numbers_iter <- function(n, acc) {
  if (n <= 0) {
    return(acc)
  } else {
    return(sum_numbers_iter(n - 1, n + acc))
  }
}

File: _experimental/_perl/numbers/lib/numbers_iterative.pl
#!/usr/bin/perl
use strict;
use warnings;

sub fibonacci {
    my ($N) = @_;
    return fibonacci_iter($N, 0, 1);
}

sub fibonacci_iter {
    my ($N, $acc2, $acc1) = @_;
    if ($N < 1) {
        return $acc2;
    } elsif ($N <= 2) {
        return $acc1 + $acc2;
    } else {
        return fibonacci_iter($N - 1, $acc1, $acc1 + $acc2);
    }
}

sub factorial {
    my ($N) = @_;
    return factorial_iter($N, 1);
}

sub factorial_iter {
    my ($N, $acc) = @_;
    if ($N <= 1) {
        return $acc;
    } else {
        return factorial_iter($N - 1, $N * $acc);
    }
}

sub sum_numbers {
    my ($N) = @_;
    return sum_numbers_iter($N, 0);
}

sub sum_numbers_iter {
    my ($N, $acc) = @_;
    if ($N <= 0) {
        return $acc;
    } else {
        return sum_numbers_iter($N - 1, $N + $acc);
    }
}

1;